core/src/sumeragi/main_loop.rs

```diff
@@ -1294,50 +1311,63 @@ where
+                                "Genesis Round Peer is revalidating the block."
+                            );
+                            let _enter = span.enter();
+                            match block.revalidate(
+                                &sumeragi.transaction_validator,
+                                &state.wsv,
+                                &state.latest_block_hash,
+                                state.latest_block_height,
+                            ) {
+                                Ok(block) => block,
+                                Err(err) => {
+                                    warn!(%err);
+                                    continue;
+                                }
+                            }
+                        };
+                        block.commit().into()
+                    }
+                    Message::BlockSyncUpdate(BlockSyncUpdate { block }) => {
+                        match block.revalidate_hashes() {
                             Ok(block) => block,
                             Err(err) => {
                                 warn!(%err);
                                 continue;
                             }
                         }
-                    };
-                    if block.header.is_genesis() && block.header.genesis_topology.is_some() {
-                        info!("Using network topology from genesis block");
-                        state.current_topology = block
-                            .header
-                            .genesis_topology
-                            .clone()
-                            .take()
-                            .expect("We just checked that it is some");
-                        commit_block(sumeragi, block, state);
-                        info!("Genesis block received and committed.");
-                        return Err(EarlyReturn::GenesisBlockReceivedAndCommitted);
                     }
-                    debug!("Received block that was not genesis.");
-                }
-                msg => {
-                    trace!(?msg, "Not handling message, waiting genesis.");
+                    msg => {
+                        trace!(?msg, "Not handling message, waiting genesis.");
+                        continue;
+                    }
+                };
+                if block.header().is_genesis() && block.header().genesis_topology.is_some() {
+                    info!("Using network topology from genesis block");
+                    state.current_topology = block
+                        .header()
+                        .genesis_topology
+                        .clone()
+                        .take()
+                        .expect("We just checked that it is some");
+                    commit_block(sumeragi, block, state);
+                    info!("Genesis block received and committed.");
+                    return Err(EarlyReturn::GenesisBlockReceivedAndCommitted);
```

Comment:
Should the success case really be an error?

Reply:
I think this was done initially to represent all ways we can exit from this function (since `()` in not reachable).
I agree that wrapping success case in `Err` looks a bit confusing.